### PR TITLE
calendar: Implement Microsoft availability and unified calendar tests

### PR DESCRIPTION
**Pull request by elie222**
# User description
calendar: Implement Microsoft availability and unified calendar tests

This PR implements Microsoft calendar availability checks and adds unified availability tests to ensure correct functionality across providers.

- Added Microsoft availability provider tests.
- Updated Microsoft availability logic to handle edge cases.
- Added comprehensive unified availability tests.
- Refactored unified availability to support new providers.

---

# Generated description

Below is a concise technical summary of the changes proposed in this PR:
```mermaid
graph LR
getUnifiedCalendarAvailability_("getUnifiedCalendarAvailability"):::modified
parseDateInTimezone_("parseDateInTimezone"):::added
fetchMicrosoftCalendarBusyPeriods_("fetchMicrosoftCalendarBusyPeriods"):::modified
MICROSOFT_GRAPH_API_("MICROSOFT_GRAPH_API"):::modified
getUnifiedCalendarAvailability_ -- "Parses start/end as calendar dates in target timezone" --> parseDateInTimezone_
fetchMicrosoftCalendarBusyPeriods_ -- "Requests UTC via Prefer header; appends 'Z' to datetimes" --> MICROSOFT_GRAPH_API_
classDef added stroke:#15AA7A
classDef removed stroke:#CD5270
classDef modified stroke:#EDAC4C
linkStyle default stroke:#CBD5E1,font-size:13px
```

Implement Microsoft calendar availability checks, ensuring correct timezone handling for busy periods and robustly processing date inputs. Refactor the <code>getUnifiedCalendarAvailability</code> function to support new providers and accurately determine day boundaries across various timezones.
<table><tr><th>Topic</th><th>Details</th><tr><td><a href=https://baz.co/changes/elie222/inbox-zero/1195?tool=ast&topic=Microsoft+Availability>Microsoft Availability</a>
        </td><td>Implement Microsoft calendar availability checks by adding a <code>Prefer: outlook.timezone="UTC"</code> header to requests and ensuring <code>dateTime</code> values from the Microsoft API are correctly formatted with a <code>Z</code> suffix for ISO 8601 compliance.<details><summary>Modified files (2)</summary><ul><li>apps/web/utils/calendar/providers/microsoft-availability.ts</li>
<li>apps/web/utils/calendar/providers/microsoft-availability.test.ts</li></ul></details><details><summary>Latest Contributors(1)</summary><table><tr><th>User</th><th>Commit</th><th>Date</th></tr><tr><td>elie222</td><td>logger</td><td>December 18, 2025</td></tr></table></details></td></tr>
<tr><td><a href=https://baz.co/changes/elie222/inbox-zero/1195?tool=ast&topic=Unified+Availability+Logic>Unified Availability Logic</a>
        </td><td>Refactor <code>getUnifiedCalendarAvailability</code> to accept <code>Date | string</code> for <code>startDate</code> and <code>endDate</code>, introducing <code>parseDateInTimezone</code> to correctly interpret calendar dates in the target timezone and prevent UTC shift issues, along with comprehensive tests for day boundary handling.<details><summary>Modified files (2)</summary><ul><li>apps/web/utils/calendar/unified-availability.ts</li>
<li>apps/web/utils/calendar/unified-availability.test.ts</li></ul></details><details><summary>Latest Contributors(1)</summary><table><tr><th>User</th><th>Commit</th><th>Date</th></tr><tr><td>elie222</td><td>logger</td><td>December 18, 2025</td></tr></table></details></td></tr></table>
This pull request is reviewed by Baz. Review like a pro on <a href=https://baz.co/changes/elie222/inbox-zero/1195?tool=ast>(Baz)</a>.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved UTC timezone handling for calendar queries to prevent conversion errors.
  * Enhanced day boundary processing to correctly handle timezone shifts.
  * Refined event filtering to exclude free and tentative events.
  * Extended support for flexible date input formats including ISO strings and YYYY-MM-DD format.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: apps/web/utils/calendar/unified-availability.ts [hunk 1]
```diff
@@ -170,3 +171,27 @@ function convertBusyPeriodsToTimezone(
...
+  dateInput: string | Date,
+  timezone: string,
+): TZDate {
+  if (dateInput instanceof Date) {
+    // For backwards compatibility: if a Date object is passed, use its UTC date components
+    // to construct the date in the target timezone
+    const year = dateInput.getUTCFullYear();
+    const month = dateInput.getUTCMonth();
+    const day = dateInput.getUTCDate();
+    return new TZDate(year, month, day, 0, 0, 0, 0, timezone);
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/calendar/unified-availability.ts:190]: Thanks for the feedback! The current implementation using UTC date components is intentional for this use case.

The function is designed to extract a "calendar date" from the Date object. The semantics are:
- For string inputs (`YYYY-MM-DD` or ISO datetime): the date is extracted directly from the string
- For Date objects: the UTC date components are used (which is why the comment says "use its UTC date components")

Using `new TZDate(dateInput, timezone)` directly would change the semantics and break existing tests. For example, `new Date("2025-11-17T00:00:00Z")` with `timezone: "America/Los_Angeles"` would interpret UTC midnight as Nov 16 at 4pm in LA, causing `startOfDay()` to return Nov 16 instead of Nov 17.

The recommended approach for callers is to pass string dates (`"2025-11-17"`) which have unambiguous semantics. The Date path exists for backwards compatibility with callers who construct UTC midnight dates.

[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/calendar/unified-availability.ts:190]: Thanks for the detailed analysis! I understand the concern about timezones ahead of UTC.

However, the suggested fix to use local components (`getFullYear/Month/Date`) would introduce a different problem: these methods use the **server's** local timezone, not the user's timezone. In a server-side function deployed on cloud infrastructure (often UTC), this would give unpredictable results depending on server location.

The current implementation using UTC components is intentional. The semantics are:
1. For **string dates** (`"2025-11-17"` or ISO datetime): date is extracted from the string directly - recommended approach
2. For **Date objects**: UTC date components are used, assuming the Date was constructed with UTC midnight

The test cases validate this behavior by passing `new Date("2025-11-17T00:00:00Z")` and expecting Nov 17.

For the edge case you mentioned (user in Australia creating a Date at local midnight), the recommended solution is to have callers pass string dates (`"2025-11-17"`) which have unambiguous semantics regardless of timezone.
